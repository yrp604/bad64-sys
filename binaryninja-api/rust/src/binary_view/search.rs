use serde_derive::{Deserialize, Serialize};

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SearchQuery {
    /// ex. "42 2e 64 65 ?? 75 67 24"
    pattern: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    start: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    end: Option<u64>,
    #[serde(rename = "ignoreCase")]
    ignore_case: bool,
    raw: bool,
    overlap: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    align: Option<u64>,
}

impl SearchQuery {
    pub fn new(pattern: impl Into<String>) -> Self {
        Self {
            pattern: pattern.into(),
            ..Default::default()
        }
    }

    /// Set the starting address for the search
    pub fn start(mut self, addr: u64) -> Self {
        self.start = Some(addr);
        self
    }

    /// Set the ending address for the search (inclusive)
    pub fn end(mut self, addr: u64) -> Self {
        self.end = Some(addr);
        self
    }

    /// Set whether to interpret the pattern as a raw string
    pub fn raw(mut self, raw: bool) -> Self {
        self.raw = raw;
        self
    }

    /// Set whether to perform case-insensitive matching
    pub fn ignore_case(mut self, ignore_case: bool) -> Self {
        self.ignore_case = ignore_case;
        self
    }

    /// Set whether to allow matches to overlap
    pub fn overlap(mut self, overlap: bool) -> Self {
        self.overlap = overlap;
        self
    }

    /// Set the alignment of matches (must be a power of 2)
    pub fn align(mut self, align: u64) -> Self {
        // Validate that align is a power of 2
        if align != 0 && (align & (align - 1)) == 0 {
            self.align = Some(align);
        }
        self
    }
}

impl SearchQuery {
    /// Serialize the query to a JSON string
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).expect("failed to serialize search query")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_search_query_builder() {
        let query = SearchQuery::new("test pattern")
            .start(0x1000)
            .end(0x2000)
            .raw(true)
            .ignore_case(true)
            .overlap(false)
            .align(16);

        assert_eq!(query.pattern, "test pattern");
        assert_eq!(query.start, Some(0x1000));
        assert_eq!(query.end, Some(0x2000));
        assert!(query.raw);
        assert!(query.ignore_case);
        assert!(!query.overlap);
        assert_eq!(query.align, Some(16));
    }

    #[test]
    fn test_search_query_json() {
        let query = SearchQuery::new("test").start(0x1000).align(8);

        let json = query.to_json();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();

        assert_eq!(parsed["pattern"], "test");
        assert_eq!(parsed["start"], 4096);
        assert_eq!(parsed["align"], 8);
        assert!(!parsed.as_object().unwrap().contains_key("end"));
    }
}
