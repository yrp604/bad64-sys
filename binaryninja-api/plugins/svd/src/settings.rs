use binaryninja::binary_view::BinaryView;
use binaryninja::settings::{QueryOptions, Settings};
use serde_json::json;
use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct LoadSettings {
    pub add_backing_regions: bool,
    pub add_bitfields: bool,
    pub add_comments: bool,
    pub auto_load_file: Option<PathBuf>,
}

impl LoadSettings {
    pub const ADD_BACKING_REGIONS_DEFAULT: bool = true;
    pub const ADD_BACKING_REGIONS_SETTING: &'static str = "analysis.svd.addBackingRegions";
    pub const ADD_BITFIELDS_DEFAULT: bool = true;
    pub const ADD_BITFIELDS_SETTING: &'static str = "analysis.svd.addBitfields";
    pub const ADD_COMMENTS_DEFAULT: bool = true;
    pub const ADD_COMMENTS_SETTING: &'static str = "analysis.svd.addComments";
    pub const AUTO_LOAD_FILE_DEFAULT: &'static str = "";
    pub const AUTO_LOAD_FILE_SETTING: &'static str = "analysis.svd.autoLoadFile";

    pub fn register() {
        let bn_settings = Settings::new();

        let add_backing_region_props = json!({
            "title" : "Add Backing Regions",
            "type" : "boolean",
            "default" : Self::ADD_BACKING_REGIONS_DEFAULT,
            "description" : "Whether to add backing regions. Backing regions allow you to write to the underlying memory of a view, but will take up space in the BNDB.",
        });
        bn_settings.register_setting_json(
            Self::ADD_BACKING_REGIONS_SETTING,
            &add_backing_region_props.to_string(),
        );

        let add_bitfields_props = json!({
            "title" : "Add Bitfields",
            "type" : "boolean",
            "default" : Self::ADD_BITFIELDS_DEFAULT,
            "description" : "Whether to add bitfields. Bitfields are not supported by Binary Ninja, so this is a workaround using unions.",
        });
        bn_settings.register_setting_json(
            Self::ADD_BITFIELDS_SETTING,
            &add_bitfields_props.to_string(),
        );

        let add_comments_props = json!({
            "title" : "Add Comments",
            "type" : "boolean",
            "default" : Self::ADD_COMMENTS_DEFAULT,
            "description" : "Whether to add comments. If you see comment placement is off, try disabling this.",
        });
        bn_settings
            .register_setting_json(Self::ADD_COMMENTS_SETTING, &add_comments_props.to_string());

        let file_props = json!({
            "title" : "SVD File",
            "type" : "string",
            "default" : Self::AUTO_LOAD_FILE_DEFAULT,
            "description" : "The SVD File to automatically load when opening the view.",
            "uiSelectionAction" : "file"
        });
        bn_settings.register_setting_json(Self::AUTO_LOAD_FILE_SETTING, &file_props.to_string());
    }

    pub fn from_view_settings(view: &BinaryView) -> Self {
        let mut load_settings = LoadSettings::default();
        let settings = Settings::new();
        let mut query_opts = QueryOptions::new_with_view(view);
        if settings.contains(Self::ADD_BACKING_REGIONS_SETTING) {
            load_settings.add_backing_regions =
                settings.get_bool_with_opts(Self::ADD_BACKING_REGIONS_SETTING, &mut query_opts);
        }
        if settings.contains(Self::ADD_BITFIELDS_SETTING) {
            load_settings.add_bitfields =
                settings.get_bool_with_opts(Self::ADD_BITFIELDS_SETTING, &mut query_opts);
        }
        if settings.contains(Self::ADD_COMMENTS_SETTING) {
            load_settings.add_comments =
                settings.get_bool_with_opts(Self::ADD_COMMENTS_SETTING, &mut query_opts);
        }
        if settings.contains(Self::AUTO_LOAD_FILE_SETTING) {
            let path_str =
                settings.get_string_with_opts(Self::AUTO_LOAD_FILE_SETTING, &mut query_opts);
            if !path_str.is_empty() {
                let path = PathBuf::from(path_str.to_string());
                load_settings.auto_load_file = Some(path);
            }
        }
        load_settings
    }
}

impl Default for LoadSettings {
    fn default() -> Self {
        Self {
            add_backing_regions: Self::ADD_BACKING_REGIONS_DEFAULT,
            add_bitfields: Self::ADD_BITFIELDS_DEFAULT,
            add_comments: Self::ADD_COMMENTS_DEFAULT,
            auto_load_file: None,
        }
    }
}
