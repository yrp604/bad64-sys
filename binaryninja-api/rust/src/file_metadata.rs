// Copyright 2021-2025 Vector 35 Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::binary_view::BinaryView;
use crate::database::Database;
use crate::rc::*;
use crate::string::*;
use binaryninjacore_sys::{
    BNBeginUndoActions, BNCloseFile, BNCommitUndoActions, BNCreateDatabase, BNCreateFileMetadata,
    BNFileMetadata, BNFileMetadataGetSessionId, BNForgetUndoActions, BNFreeFileMetadata,
    BNGetCurrentOffset, BNGetCurrentView, BNGetExistingViews, BNGetFileMetadataDatabase,
    BNGetFileViewOfType, BNGetFilename, BNGetProjectFile, BNIsAnalysisChanged,
    BNIsBackedByDatabase, BNIsFileModified, BNMarkFileModified, BNMarkFileSaved, BNNavigate,
    BNNewFileReference, BNOpenDatabaseForConfiguration, BNOpenExistingDatabase, BNRedo,
    BNRevertUndoActions, BNSaveAutoSnapshot, BNSetFilename, BNUndo,
};
use binaryninjacore_sys::{BNCreateDatabaseWithProgress, BNOpenExistingDatabaseWithProgress};
use std::ffi::c_void;
use std::fmt::Debug;
use std::path::Path;

use crate::progress::ProgressCallback;
use crate::project::file::ProjectFile;
use std::ptr::{self, NonNull};

#[derive(PartialEq, Eq, Hash)]
pub struct FileMetadata {
    pub(crate) handle: *mut BNFileMetadata,
}

impl FileMetadata {
    pub(crate) fn from_raw(handle: *mut BNFileMetadata) -> Self {
        Self { handle }
    }

    pub(crate) fn ref_from_raw(handle: *mut BNFileMetadata) -> Ref<Self> {
        unsafe { Ref::new(Self { handle }) }
    }

    pub fn new() -> Ref<Self> {
        Self::ref_from_raw(unsafe { BNCreateFileMetadata() })
    }

    pub fn with_filename(name: &str) -> Ref<Self> {
        let ret = FileMetadata::new();
        ret.set_filename(name);
        ret
    }

    pub fn close(&self) {
        unsafe {
            BNCloseFile(self.handle);
        }
    }

    pub fn session_id(&self) -> usize {
        unsafe { BNFileMetadataGetSessionId(self.handle) }
    }

    pub fn filename(&self) -> String {
        unsafe {
            let raw = BNGetFilename(self.handle);
            BnString::into_string(raw)
        }
    }

    pub fn set_filename(&self, name: &str) {
        let name = name.to_cstr();

        unsafe {
            BNSetFilename(self.handle, name.as_ptr());
        }
    }

    pub fn modified(&self) -> bool {
        unsafe { BNIsFileModified(self.handle) }
    }

    pub fn mark_modified(&self) {
        unsafe {
            BNMarkFileModified(self.handle);
        }
    }

    pub fn mark_saved(&self) {
        unsafe {
            BNMarkFileSaved(self.handle);
        }
    }

    pub fn is_analysis_changed(&self) -> bool {
        unsafe { BNIsAnalysisChanged(self.handle) }
    }

    pub fn is_database_backed(&self) -> bool {
        self.is_database_backed_for_view_type("")
    }

    pub fn is_database_backed_for_view_type(&self, view_type: &str) -> bool {
        let view_type = view_type.to_cstr();

        unsafe { BNIsBackedByDatabase(self.handle, view_type.as_ref().as_ptr() as *const _) }
    }

    /// Runs a failable function where the failure state will revert any undo actions that occurred
    /// during the time of the function's execution.
    ///
    /// NOTE: This will commit or undo any actions that occurred on **any** thread as this state is not thread local.
    ///
    /// NOTE: This is **NOT** thread safe, if you are holding any locks that might be held by both the main thread
    /// and the thread executing this function, you can deadlock. You should also never call this function
    /// on multiple threads at a time. See the following issues:
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6289>
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6325>
    pub fn run_undoable_transaction<F: FnOnce() -> Result<T, E>, T, E>(
        &self,
        func: F,
    ) -> Result<T, E> {
        let undo = self.begin_undo_actions(false);
        let result = func();
        match result {
            Ok(t) => {
                self.commit_undo_actions(&undo);
                Ok(t)
            }
            Err(e) => {
                self.revert_undo_actions(&undo);
                Err(e)
            }
        }
    }

    /// Creates a new undo entry, any undo actions after this will be added to this entry.
    ///
    /// NOTE: This is **NOT** thread safe, if you are holding any locks that might be held by both the main thread
    /// and the thread executing this function, you can deadlock. You should also never call this function
    /// on multiple threads at a time. See the following issues:
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6289>
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6325>
    pub fn begin_undo_actions(&self, anonymous_allowed: bool) -> String {
        unsafe { BnString::into_string(BNBeginUndoActions(self.handle, anonymous_allowed)) }
    }

    /// Commits the undo entry with the id to the undo buffer.
    ///
    /// NOTE: This is **NOT** thread safe, if you are holding any locks that might be held by both the main thread
    /// and the thread executing this function, you can deadlock. You should also never call this function
    /// on multiple threads at a time. See the following issues:
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6289>
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6325>
    pub fn commit_undo_actions(&self, id: &str) {
        let id = id.to_cstr();
        unsafe {
            BNCommitUndoActions(self.handle, id.as_ref().as_ptr() as *const _);
        }
    }

    /// Reverts the undo actions committed in the undo entry.
    ///
    /// NOTE: This is **NOT** thread safe, if you are holding any locks that might be held by both the main thread
    /// and the thread executing this function, you can deadlock. You should also never call this function
    /// on multiple threads at a time. See the following issues:
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6289>
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6325>
    pub fn revert_undo_actions(&self, id: &str) {
        let id = id.to_cstr();
        unsafe {
            BNRevertUndoActions(self.handle, id.as_ref().as_ptr() as *const _);
        }
    }

    /// Forgets the undo actions committed in the undo entry.
    ///
    /// NOTE: This is **NOT** thread safe, if you are holding any locks that might be held by both the main thread
    /// and the thread executing this function, you can deadlock. You should also never call this function
    /// on multiple threads at a time. See the following issues:
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6289>
    ///  - <https://github.com/Vector35/binaryninja-api/issues/6325>
    pub fn forget_undo_actions(&self, id: &str) {
        let id = id.to_cstr();
        unsafe {
            BNForgetUndoActions(self.handle, id.as_ref().as_ptr() as *const _);
        }
    }

    pub fn undo(&self) {
        unsafe {
            BNUndo(self.handle);
        }
    }

    pub fn redo(&self) {
        unsafe {
            BNRedo(self.handle);
        }
    }

    pub fn current_view(&self) -> String {
        unsafe { BnString::into_string(BNGetCurrentView(self.handle)) }
    }

    pub fn current_offset(&self) -> u64 {
        unsafe { BNGetCurrentOffset(self.handle) }
    }

    /// Navigate to an offset for a specific view.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use binaryninja::file_metadata::FileMetadata;
    /// # let file: FileMetadata = unimplemented!();
    /// file.navigate_to("Linear:Raw", 0x0).expect("Linear:Raw should always be present");
    /// ```
    pub fn navigate_to(&self, view: &str, offset: u64) -> Result<(), ()> {
        let view = view.to_cstr();

        unsafe {
            if BNNavigate(self.handle, view.as_ref().as_ptr() as *const _, offset) {
                Ok(())
            } else {
                Err(())
            }
        }
    }

    /// Get the [`BinaryView`] for the view type.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use binaryninja::file_metadata::FileMetadata;
    /// # let file: FileMetadata = unimplemented!();
    /// file.view_of_type("Raw").expect("Raw type should always be present");
    /// ```
    pub fn view_of_type(&self, view: &str) -> Option<Ref<BinaryView>> {
        let view = view.to_cstr();

        unsafe {
            let raw_view_ptr = BNGetFileViewOfType(self.handle, view.as_ref().as_ptr() as *const _);
            match raw_view_ptr.is_null() {
                false => Some(BinaryView::ref_from_raw(raw_view_ptr)),
                true => None,
            }
        }
    }

    pub fn view_types(&self) -> Array<BnString> {
        let mut count = 0;
        unsafe {
            let types = BNGetExistingViews(self.handle, &mut count);
            Array::new(types, count, ())
        }
    }

    /// Get the [`ProjectFile`] for the [`FileMetadata`].
    pub fn project_file(&self) -> Option<Ref<ProjectFile>> {
        unsafe {
            let res = NonNull::new(BNGetProjectFile(self.handle))?;
            Some(ProjectFile::ref_from_raw(res))
        }
    }

    pub fn create_database(&self, file_path: impl AsRef<Path>) -> bool {
        // Databases are created with the root view (Raw).
        let Some(raw_view) = self.view_of_type("Raw") else {
            return false;
        };

        let file_path = file_path.as_ref().to_cstr();
        unsafe {
            BNCreateDatabase(
                raw_view.handle,
                file_path.as_ptr() as *mut _,
                ptr::null_mut(),
            )
        }
    }

    // TODO: Pass settings?
    pub fn create_database_with_progress<P: ProgressCallback>(
        &self,
        file_path: impl AsRef<Path>,
        mut progress: P,
    ) -> bool {
        // Databases are created with the root view (Raw).
        let Some(raw_view) = self.view_of_type("Raw") else {
            return false;
        };
        let file_path = file_path.as_ref().to_cstr();
        unsafe {
            BNCreateDatabaseWithProgress(
                raw_view.handle,
                file_path.as_ptr() as *mut _,
                &mut progress as *mut P as *mut c_void,
                Some(P::cb_progress_callback),
                ptr::null_mut(),
            )
        }
    }

    pub fn save_auto_snapshot(&self) -> bool {
        // Snapshots are saved with the root view (Raw).
        let Some(raw_view) = self.view_of_type("Raw") else {
            return false;
        };

        unsafe { BNSaveAutoSnapshot(raw_view.handle, ptr::null_mut() as *mut _) }
    }

    pub fn open_database_for_configuration(&self, file: &Path) -> Result<Ref<BinaryView>, ()> {
        let file = file.to_cstr();
        unsafe {
            let bv =
                BNOpenDatabaseForConfiguration(self.handle, file.as_ref().as_ptr() as *const _);

            if bv.is_null() {
                Err(())
            } else {
                Ok(BinaryView::ref_from_raw(bv))
            }
        }
    }

    pub fn open_database(&self, file: &Path) -> Result<Ref<BinaryView>, ()> {
        let file = file.to_cstr();
        let view = unsafe { BNOpenExistingDatabase(self.handle, file.as_ptr()) };

        if view.is_null() {
            Err(())
        } else {
            Ok(unsafe { BinaryView::ref_from_raw(view) })
        }
    }

    pub fn open_database_with_progress<P: ProgressCallback>(
        &self,
        file: &Path,
        mut progress: P,
    ) -> Result<Ref<BinaryView>, ()> {
        let file = file.to_cstr();

        let view = unsafe {
            BNOpenExistingDatabaseWithProgress(
                self.handle,
                file.as_ptr(),
                &mut progress as *mut P as *mut c_void,
                Some(P::cb_progress_callback),
            )
        };

        if view.is_null() {
            Err(())
        } else {
            Ok(unsafe { BinaryView::ref_from_raw(view) })
        }
    }

    /// Get the current database
    pub fn database(&self) -> Option<Ref<Database>> {
        let result = unsafe { BNGetFileMetadataDatabase(self.handle) };
        NonNull::new(result).map(|handle| unsafe { Database::ref_from_raw(handle) })
    }
}

impl Debug for FileMetadata {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("FileMetadata")
            .field("filename", &self.filename())
            .field("session_id", &self.session_id())
            .field("modified", &self.modified())
            .field("is_analysis_changed", &self.is_analysis_changed())
            .field("current_view_type", &self.current_view())
            .field("current_offset", &self.current_offset())
            .field("view_types", &self.view_types().to_vec())
            .finish()
    }
}

unsafe impl Send for FileMetadata {}
unsafe impl Sync for FileMetadata {}

impl ToOwned for FileMetadata {
    type Owned = Ref<Self>;

    fn to_owned(&self) -> Self::Owned {
        unsafe { RefCountable::inc_ref(self) }
    }
}

unsafe impl RefCountable for FileMetadata {
    unsafe fn inc_ref(handle: &Self) -> Ref<Self> {
        Ref::new(Self {
            handle: BNNewFileReference(handle.handle),
        })
    }

    unsafe fn dec_ref(handle: &Self) {
        BNFreeFileMetadata(handle.handle);
    }
}
