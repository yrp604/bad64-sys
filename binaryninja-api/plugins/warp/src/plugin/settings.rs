use binaryninja::settings::{QueryOptions, Settings as BNSettings};
use serde_json::json;
use std::string::ToString;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PluginSettings {
    /// Whether to load bundled WARP files on startup. Turn this off if you want to manually load them.
    ///
    /// This is set to [PluginSettings::LOAD_BUNDLED_FILES_DEFAULT] by default.
    pub load_bundled_files: bool,
    /// Whether to load user WARP files on startup. Turn this off if you want to manually load them.
    ///
    /// This is set to [PluginSettings::LOAD_USER_FILES_DEFAULT] by default.
    pub load_user_files: bool,
    /// The WARP server to use.
    ///
    /// This is set to [PluginSettings::SERVER_URL_DEFAULT] by default.
    pub server_url: String,
    /// The API key to use for the selected WARP server, if not specified, you will be unable to push data and may be rate-limited.
    ///
    /// This is set to [PluginSettings::SERVER_API_KEY_DEFAULT] by default.
    pub server_api_key: Option<String>,
    pub second_server_url: Option<String>,
    pub second_server_api_key: Option<String>,
    /// A source must have at least one of these tags to be considered a valid source.
    ///
    /// This is set to [PluginSettings::SOURCE_TAGS_DEFAULT] by default.
    pub whitelisted_source_tags: Vec<String>,
    /// Whether to allow networked WARP requests. Turning this off will not disable local WARP functionality.
    ///
    /// This is set to [PluginSettings::ENABLE_SERVER_DEFAULT] by default.
    pub enable_server: bool,
}

impl PluginSettings {
    pub const WHITELISTED_SOURCE_TAGS_DEFAULT: Vec<String> = vec![];
    pub const WHITELISTED_SOURCE_TAGS_SETTING: &'static str = "analysis.warp.whitelistedSourceTags";
    pub const LOAD_BUNDLED_FILES_DEFAULT: bool = true;
    pub const LOAD_BUNDLED_FILES_SETTING: &'static str = "analysis.warp.loadBundledFiles";
    pub const LOAD_USER_FILES_DEFAULT: bool = true;
    pub const LOAD_USER_FILES_SETTING: &'static str = "analysis.warp.loadUserFiles";
    pub const SERVER_URL_DEFAULT: &'static str = "https://warp.binary.ninja";
    pub const SERVER_URL_SETTING: &'static str = "analysis.warp.serverUrl";
    pub const SERVER_API_KEY_DEFAULT: Option<String> = None;
    pub const SERVER_API_KEY_SETTING: &'static str = "analysis.warp.serverApiKey";
    pub const SECONDARY_SERVER_URL_DEFAULT: Option<String> = None;
    pub const SECONDARY_SERVER_URL_SETTING: &'static str = "analysis.warp.secondServerUrl";
    pub const SECONDARY_SERVER_API_KEY_DEFAULT: Option<String> = None;
    pub const SECONDARY_SERVER_API_KEY_SETTING: &'static str = "analysis.warp.secondServerApiKey";
    pub const ENABLE_SERVER_DEFAULT: bool = false;
    pub const ENABLE_SERVER_SETTING: &'static str = "network.enableWARP";

    pub fn register(bn_settings: &mut BNSettings) {
        let whitelisted_source_tags_prop = json!({
            "title" : "Blacklisted Sources",
            "type" : "array",
            "default" : Self::WHITELISTED_SOURCE_TAGS_DEFAULT,
            "description" : "Add a sources UUID to this list to blacklist it from being considered a valid source. This is useful for sources that are known to be false positives.",
            "ignore" : [],
        });
        bn_settings.register_setting_json(
            Self::WHITELISTED_SOURCE_TAGS_SETTING,
            &whitelisted_source_tags_prop.to_string(),
        );
        let load_bundled_files_prop = json!({
            "title" : "Load Bundled Files",
            "type" : "boolean",
            "default" : Self::LOAD_BUNDLED_FILES_DEFAULT,
            "description" : "Whether to load bundled WARP files on startup. Turn this off if you want to manually load them.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::LOAD_BUNDLED_FILES_SETTING,
            &load_bundled_files_prop.to_string(),
        );
        let load_user_files_prop = json!({
            "title" : "Load User Files",
            "type" : "boolean",
            "default" : Self::LOAD_USER_FILES_DEFAULT,
            "description" : "Whether to load user WARP files on startup. Turn this off if you want to manually load them.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::LOAD_USER_FILES_SETTING,
            &load_user_files_prop.to_string(),
        );
        let server_url_prop = json!({
            "title" : "Server URL",
            "type" : "string",
            "default" : Self::SERVER_URL_DEFAULT,
            "description" : "The WARP server to use.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(Self::SERVER_URL_SETTING, &server_url_prop.to_string());
        let server_api_key_prop = json!({
            "title" : "Server API Key",
            "type" : "string",
            "default" : Self::SERVER_API_KEY_DEFAULT,
            "description" : "The API key to use for the selected WARP server, if not specified you will be unable to push data, and may be rate limited.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "hidden": true,
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::SERVER_API_KEY_SETTING,
            &server_api_key_prop.to_string(),
        );
        let second_server_url_prop = json!({
            "title" : "Secondary Server URL",
            "type" : "string",
            "default" : Self::SECONDARY_SERVER_URL_DEFAULT,
            "description" : "",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::SECONDARY_SERVER_URL_SETTING,
            &second_server_url_prop.to_string(),
        );
        let second_server_api_key_prop = json!({
            "title" : "Secondary Server API Key",
            "type" : "string",
            "default" : Self::SECONDARY_SERVER_API_KEY_DEFAULT,
            "description" : "",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "hidden": true,
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::SECONDARY_SERVER_API_KEY_SETTING,
            &second_server_api_key_prop.to_string(),
        );
        let server_enabled_prop = json!({
            "title" : "Enable WARP",
            "type" : "boolean",
            "default" : Self::ENABLE_SERVER_DEFAULT,
            "description" : "Whether or not to allow networked WARP requests. Turning this off will not disable local WARP functionality.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::ENABLE_SERVER_SETTING,
            &server_enabled_prop.to_string(),
        );
    }

    /// Retrieve plugin settings from [`BNSettings`].
    pub fn from_settings(bn_settings: &BNSettings, query_opts: &mut QueryOptions) -> Self {
        let mut settings = PluginSettings::default();
        if bn_settings.contains(Self::LOAD_BUNDLED_FILES_SETTING) {
            settings.load_bundled_files = bn_settings.get_bool(Self::LOAD_BUNDLED_FILES_SETTING);
        }
        if bn_settings.contains(Self::LOAD_USER_FILES_SETTING) {
            settings.load_user_files = bn_settings.get_bool(Self::LOAD_USER_FILES_SETTING);
        }
        if bn_settings.contains(Self::SERVER_URL_SETTING) {
            settings.server_url = bn_settings.get_string(Self::SERVER_URL_SETTING);
        }
        if bn_settings.contains(Self::SERVER_API_KEY_SETTING) {
            let server_api_key_str = bn_settings.get_string(Self::SERVER_API_KEY_SETTING);
            if !server_api_key_str.is_empty() {
                settings.server_api_key = Some(server_api_key_str);
            }
        }
        if bn_settings.contains(Self::SECONDARY_SERVER_URL_SETTING) {
            let server_api_key_str = bn_settings.get_string(Self::SECONDARY_SERVER_URL_SETTING);
            if !server_api_key_str.is_empty() {
                settings.second_server_url = Some(server_api_key_str);
            }
        }
        if bn_settings.contains(Self::SECONDARY_SERVER_API_KEY_SETTING) {
            let server_api_key_str = bn_settings.get_string(Self::SECONDARY_SERVER_API_KEY_SETTING);
            if !server_api_key_str.is_empty() {
                settings.second_server_api_key = Some(server_api_key_str);
            }
        }
        if bn_settings.contains(Self::ENABLE_SERVER_SETTING) {
            settings.enable_server = bn_settings.get_bool(Self::ENABLE_SERVER_SETTING);
        }

        if bn_settings.contains(Self::WHITELISTED_SOURCE_TAGS_SETTING) {
            let whitelisted_source_tags_str = bn_settings
                .get_string_list_with_opts(Self::WHITELISTED_SOURCE_TAGS_SETTING, query_opts);
            settings.whitelisted_source_tags = whitelisted_source_tags_str
                .iter()
                .map(|s| s.to_string())
                .collect();
        }
        settings
    }
}

impl Default for PluginSettings {
    fn default() -> Self {
        Self {
            whitelisted_source_tags: PluginSettings::WHITELISTED_SOURCE_TAGS_DEFAULT,
            load_bundled_files: PluginSettings::LOAD_BUNDLED_FILES_DEFAULT,
            load_user_files: PluginSettings::LOAD_USER_FILES_DEFAULT,
            server_url: PluginSettings::SERVER_URL_DEFAULT.to_string(),
            server_api_key: PluginSettings::SERVER_API_KEY_DEFAULT,
            second_server_url: PluginSettings::SECONDARY_SERVER_URL_DEFAULT,
            second_server_api_key: PluginSettings::SECONDARY_SERVER_API_KEY_DEFAULT,
            enable_server: PluginSettings::ENABLE_SERVER_DEFAULT,
        }
    }
}
