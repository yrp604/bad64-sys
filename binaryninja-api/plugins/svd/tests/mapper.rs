use binaryninja::binary_view::BinaryViewExt;
use binaryninja::binary_view::{BinaryView, BinaryViewBase};
use binaryninja::file_metadata::FileMetadata;
use binaryninja::headless::Session;
use std::path::PathBuf;
use svd_ninja::mapper::DeviceMapper;
use svd_ninja::settings::LoadSettings;

// These are the target files present in OUT_DIR
// Add the files to fixtures
static TARGET_FILES: [&str; 2] = ["ARM_Sample.svd", "esp32c2.svd"];

#[test]
fn insta_snapshots() {
    let _session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let svd_str = std::fs::read_to_string(&path).expect("Failed to read svd file");
        let device = svd_parser::parse(&svd_str).expect("Failed to parse svd file");
        let view = BinaryView::from_data(&FileMetadata::new(), &[]).expect("Failed to create view");
        let address_size = view.address_size();
        DeviceMapper::new(LoadSettings::default(), address_size, device).map_to_view(&view);

        let types = view
            .types()
            .to_vec()
            .iter()
            .map(|ty| ty.ty.to_string())
            .collect::<Vec<String>>();
        let types_snapshot_name = format!("{}_types", file_name);
        insta::assert_debug_snapshot!(types_snapshot_name, types);

        // TODO: This does not work yet because memory map is not turned on for raw view?
        // let value: Value = serde_json::from_str(&view.memory_map().description()).unwrap();
        // let map = view.memory_map().description();
        // let memory_snapshot_name = format!("{}_memory_map", file_name);
        // insta::assert_debug_snapshot!(memory_snapshot_name, value);
    }
}

#[test]
fn test_bitfield_unions() {
    let _session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    let path = out_dir.join("ARM_Sample.svd");
    let svd_str = std::fs::read_to_string(&path).expect("Failed to read svd file");
    let device = svd_parser::parse(&svd_str).expect("Failed to parse svd file");
    let view = BinaryView::from_data(&FileMetadata::new(), &[]).expect("Failed to create view");
    let address_size = view.address_size();
    let mapper = DeviceMapper::new(LoadSettings::default(), address_size, device.clone());

    let peripheral = device.get_peripheral("TIMER0").unwrap();
    let register = peripheral.get_register("SR").unwrap();
    let register_ty = mapper.register_type(&register);
    insta::assert_debug_snapshot!("bitfield_union_types", register_ty);
}
