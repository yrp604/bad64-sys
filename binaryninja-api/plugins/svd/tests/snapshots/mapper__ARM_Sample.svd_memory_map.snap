---
source: plugins/svd/tests/mapper.rs
expression: value
---
Object {
    "MemoryMap": Array [],
}
